enhancements/proposed/20200501-json-output/start-json-output.md

```diff
@@ -0,0 +1,231 @@
+{"Name":"Selecting Driver","Message":"Using the hyperkit driver,"TotalSteps":9,"CurrentStep":2,"Type":"Log"}
+{"Name":"Starting Control Plane","Message":"Starting node minikube in cluster minikube","TotalSteps":9,"CurrentStep":3,"Type":"Log"}
+{"Name":"Download Necessary Artifacts","Message":"Downloading Kubernetes v1.18.1 preload","TotalSteps":9,"CurrentStep":4,"Type":"Log"}
+  {"Type":"Download", "Artifact":"preload.tar.gz", "Progress": "10%"}
+  {"Type":"Download", "Artifact":"preload.tar.gz", "Progress": "61%"}
+  {"Type":"Download", "Artifact":"preload.tar.gz", "Progress": "73%"}
+  {"Type":"Download", "Artifact":"preload.tar.gz", "Progress": "87%"}
+  {"Type":"Download", "Artifact":"preload.tar.gz", "Progress": "100%"}
+{"Name":"Creating Node","Message":"Creating hyperkit VM","TotalSteps":9,"CurrentStep":5,"Type":"Log"}
+{"Name":"Preparing Kubernetes","Message":"Preparing Kubernetes v1.18.1 on Docker 19.03.8","TotalSteps":9,"CurrentStep":6,"Type":"Log"}
+{"Name":"Verifying Kubernetes","Message":"🔎  Verifying Kubernetes components","TotalSteps":9,"CurrentStep":7,"Type":"Log"}
+{"Name":"Enabling Addons","Message":"🌟  Enabled addons: default-storageclass, storage-provisioner","TotalSteps":9,"CurrentStep":7,"Type":"Log"}
+{"Name":"Done","Message":"🏄  Done! kubectl is now configured to use \"minikube\"","TotalSteps":9,"CurrentStep":9,"Type":"Log"}
+```
+
+This way, clients can parse the output as it is logged and know the following:
+1. What the type of step is (Log vs Download)
+1. What step we are currently on
+1. The total number of steps
+1. The specific message related to that step
+
+
+## Implementation Details
+
+### Stderr
+glog logs can be sent to stderr as usual.
+In addition, we will output error code from [err_map.go](https://github.com/kubernetes/minikube/blob/master/pkg/minikube/problem/err_map.go) as a parsable JSON message to stderr.
+
+This will be done by adding the following function to the `out` package:
+
+```go
+func DisplayErrorJSON(out io.Writer, p *problem.Problem)
+```
+
+which will print the JSON encoding of `p` to `out` (in this case, stderr).
+
+
+### Stdout - Log Steps
+Since we need to approximate the total number of steps before minikube starts, we need to know the general steps we expect to execute before starting.
+
+I propose creating a registry of logs, which is prefilled with the following steps:
+
+* "Minikube Version"
+* "Selecting Driver"
+*	"Starting Control Plane"
+*	"Download Necessary Artifacts"
+*	"Creating Node"
+*	"Preparing Kubernetes"
+*	"Verifying Kubernetes"
+*	"Enabling Addons"
+*	"Done"
+
+When a log is called in the code, it will be associated with one of the above steps.
+
+This will allow us to determine which number step we are currently on, and include that information in the output.
+
+_Note: We may skip steps depending on the user's existing setup. For example, we may not need to "Verify Kubernetes" on a soft start. The current step/total step numbers will only be approximations_
```

Comment:
Can you bound those approximations? I'm concerned that if minikube only "sometimes" increases the number of steps, that's likely to be missed by consumers. Then, their minikube progress bar is glitchy (e.g. sometimes draws past 100%), and minikube ends up looking buggy.

Is it possible to include all possible steps in the count, so that TotalSteps never changes but CurrentStep often skips some numbers? That seems more robust against mistakes by event consumers. 

If not, i.e. if an unknown number of steps could suddenly get scheduled after the run starts, consider s/TotalSteps/EstimatedSteps/ or consider putting those new steps as fractional steps somewhere so progress bars continue to work.

Reply:
For sure, total steps should never change. Current steps may skip ahead, but it seems like that's ok. I'll rewrite this so it's clearer.